internal/subscriber/subscriber.go

```diff
@@ -87,78 +76,48 @@ func (s *Subscriber) GetID() string {
+func (s *Subscriber) retry(ctx context.Context, msg messagebroker.ReceivedMessage) {
+
+	// for older subscriptions, delayConfig will not get auto-created
+	if s.retrier == nil {
+		logger.Ctx(ctx).Infow("subscriber: skipping retry as retrier not configured", "logFields", s.getLogFields())
 		return
 	}
 
-	logFields := s.getLogFields()
-	logFields["messageId"] = retryMsg.MessageID
-
-	startTime := time.Now()
-
-	// remove message from the primary topic
-	_, err := s.consumer.CommitByPartitionAndOffset(ctx, messagebroker.CommitOnTopicRequest{
-		Topic:     retryMsg.Topic,
-		Partition: retryMsg.Partition,
-		// add 1 to current offset
-		// https://docs.confluent.io/5.5.0/clients/confluent-kafka-go/index.html#pkg-overview
-		Offset: retryMsg.Offset + 1,
-	})
-
+	// prepare message headers to be used by retrier
+	msg.SourceTopic = s.subscription.Topic
+	msg.RetryTopic = s.subscription.GetRetryTopic() // push retried messages to primary retry topic
+	msg.CurrentTopic = s.subscription.Topic         // initially these will be same
+	msg.Subscription = s.subscription.DelayConfig.Subscription
+	msg.CurrentRetryCount = msg.CurrentRetryCount + 1 // should be zero to begin with
+	msg.MaxRetryCount = s.subscription.DelayConfig.MaxDeliveryAttempts
+	msg.DeadLetterTopic = s.subscription.DelayConfig.DeadLetterTopic
+	msg.InitialDelayInterval = s.subscription.DelayConfig.MinimumBackoffInSeconds
+
+	err := s.retrier.Handle(ctx, msg)
 	if err != nil {
-		logFields["error"] = err.Error()
-		logger.Ctx(ctx).Errorw("subscriber: commit to primary topic failed", "logFields", logFields)
+		logger.Ctx(ctx).Errorw("subscriber: push to retrier failed", "logFields", s.getLogFields(), "error", err.Error())
 		s.errChan <- err
 		return
 	}
 
-	// check max retries.
-	if retryMsg.RetryCount >= maxMessageRetryAttempts {
-		// then push message to the dlq topic
-		logger.Ctx(ctx).Infow("subscriber: max retries exceeded. pushing to dlq topic", "logFields", logFields)
-		_, err = s.dlqProducer.SendMessage(ctx, messagebroker.SendMessageToTopicRequest{
-			Topic:     s.dlqTopic,
-			Message:   retryMsg.Data,
-			TimeoutMs: 50,
-		})
-
-		if err != nil {
-			logFields["error"] = err.Error()
-			logger.Ctx(ctx).Errorw("subscriber: push to dlq topic failed", "logFields", logFields)
-			s.errChan <- err
-			return
-		}
-
-		subscriberMessagesRetried.WithLabelValues(env, s.dlqTopic, s.subscription).Inc()
-		logger.Ctx(ctx).Infow("subscriber: msg pushed to dlq topic", "logFields", logFields)
-		return
-	}
-
-	// then push message to the retry topic
-	_, err = s.retryProducer.SendMessage(ctx, messagebroker.SendMessageToTopicRequest{
-		Topic:      s.retryTopic,
-		Message:    retryMsg.Data,
-		TimeoutMs:  50,
-		MessageID:  retryMsg.MessageID,
-		RetryCount: retryMsg.incrementAndGetRetryCount(),
+	// commit on the primary topic after message has been submitted for retry
+	_, err = s.consumer.CommitByPartitionAndOffset(ctx, messagebroker.CommitOnTopicRequest{
```

Comment:
if you directly commit this, how does your in memory map updated, shouldn't commit happen as part of acknowledge ?

Reply:
For every `s.retry()` call, we call a subsequent `s.removeMessageFromMemory(ctx, stats, req.MessageID)` from the calling function. This takes care of cleaning up the in-memory ds.